### DISCUSSION TITLE
Unable to create objects owned by test_schema with Ingres dialect

### DISCUSSION THREAD

**Discussion by hab6** [sqlalchemy/sqlalchemy#11366] [Usage Questions]
I am finding that many tests in the [dialect compliance suite](https://github.com/sqlalchemy/sqlalchemy/tree/main/lib/sqlalchemy/testing/suite) are failing when using the [Ingres dialect](https://github.com/ActianCorp/sqlalchemy-ingres) because Ingres does not allow users to create and drop tables under different schemas/usernames and certain tests use a schema name of `test_schema`.
 
#### Example Error
    File "C:\work\sqlalchemy\lib\sqlalchemy\engine\default.py", line 919,
    in do_execute cursor.execute(statement, parameters)
    sqlalchemy.exc.ProgrammingError: (pyodbc.ProgrammingError) ('42503', "[42503] [Actian][Actian II ODBC Driver] 
    [INGRES]CREATE TABLE: You may not create an object owned by 'test_schema'. (328737) (SQLExecDirectW)")
    [SQL:
    CREATE TABLE test_schema.test_table (
        id INTEGER GENERATED BY DEFAULT AS IDENTITY NOT NULL,
        data VARCHAR(50),
        PRIMARY KEY (id) ) ]

I think in _some_ cases the error could be avoided by running the tests simply as user `test_schema`, but there are other tests such as `test_select.py:class SameNamedSchemaTableTest(fixtures.TablesTest)` that explicitly test operations under multiple schemas.

Referencing SQLAlchemy [README.unittests.rst](https://github.com/sqlalchemy/sqlalchemy/blob/main/README.unittests.rst#database-configuration), it instructs that "Several tests require alternate usernames or schemas to be present", referring to `test_schema` (and `test_schema_2` for PostgreSQL and MSSQL only) and that "the database user must have permission to create and drop tables within these schemas".

With Ingres, the only way to create objects under a different schema is if the current Ingres user has **`security administrator`** and **`operator`** privileges and then must execute the command `SET SESSION AUTHORIZATION username` in order to set the effective user for the session and then is able to create and drop objects under the new user.

Here is an example of using `SET SESSION AUTHORIZATION ...` with Ingres to create tables with the same name under different usersnames.

    set session authorization user1\g
    create table simptab (i integer)\g
    commit\g
    set session authorization user2\g
    create table simptab (i integer)\g
    commit\g
    set session authorization user3\g
    create table simptab (i integer)\g
    commit\g
    set session authorization user4\g
    create table simptab (i integer)\g
    commit\g

    ===== The following tables would subsequently exist =====

    schema    table
    --------  -------
    user1     simptab
    user2     simptab
    user3     simptab
    user4     simptab

#### Assistance Request
I appreciate any suggestions on how to get this to work with the Ingres restriction.

I thought maybe using a custom `provision.py` (from `sqlalchemy.testing.provision`) and `requirements.py` (from `sqlalchemy.testing.requirements`) could somehow be used to run the `SET SESSION AUTHORIZATION ...` command prior to creating/dropping objects by first setting the appropriate user from the schema specified in the object definition, for example:

    def define_tables(cls, metadata):
        Table(
            "some_table",
            metadata,
            Column("id", Integer, primary_key=True),
            schema=config.test_schema,
        )
                              
But I am unable to get this work.  Thanks for any advice.

#### Environment
    Microsoft Windows [Version 10.0.19045.4170]
    DBMS: Ingres 11.2
    Python 3.10.7
    mock              5.1.0
    packaging         24.0
    pip               24.0
    pluggy            1.5.0
    psycopg           3.1.18
    psycopg2          2.9.9
    pyodbc            5.1.0
    pyproject-api     1.6.1
    pypyodbc          1.3.6
    pytest            8.2.0
    setuptools        63.2.0
    SQLAlchemy        2.0.29.dev0
    sqlalchemy-ingres 0.0.7.dev3
    tox               4.15.0

**Comment by zzzeek** [comment]
your custom dialect should include a class-level boolean `supports_schemas=False`, this seems to be missing from the base Dialect class but it's part of DefaultDialect [here](https://github.com/sqlalchemy/sqlalchemy/blob/02001e9458802ebb512a140aa24e663b364dc3ad/lib/sqlalchemy/engine/default.py#L145).  This then drives the requirements.py rule [here](https://github.com/sqlalchemy/sqlalchemy/blob/02001e9458802ebb512a140aa24e663b364dc3ad/lib/sqlalchemy/testing/requirements.py#L543) which you could also override manually (See requirements.py at https://github.com/sqlalchemy/sqlalchemy/blob/main/README.dialects.rst#dialect-layout)

**Reply by hab6** [reply]
Thanks for the tip about setting `supports_schemas=False` which indeed gives relief by skipping the related tests.

I realized that it probably doesn't make sense to go down the path of trying to get the Ingres dialect to run the suite tests that create tables under the schema (`test_schema`) which is different from the current user since Ingres doesn't support that kind of thing without the user having unusually elevated privileges which are required to run `SET SESSION AUTHORIZATION ...` to change the effective user for the session.
